### DISCUSSION TITLE
Support for non-cacheable requests in gem5 CHI protocol

### DISCUSSION THREAD

**Discussion by hnpl** [gem5/gem5#815] [Q&A] [answered]
Hi,

I wanted to ask whether the gem5's implementation of the CHI protocol support non-cacheable/un-cacheable requests from the core. I ran an experiment where a linear traffic generator produces a series of un-cacheable writes to the CHI cache system; however, the CHI cache cached the writes. Is it an expected behavior?

More details on the experiment: the destination of the un-cacheable writes are physical pages in memory, not MMIOs to an I/O device. I checked and saw the the memory packets, which were seen by the CHI cache, had the un-cacheable flag set.

**Comment by powerjg** [comment]
Tagging @tiagormk

**Reply by tiagormk** [reply, marked as answer]
Hi @hnpl, all accesses to physical memory that go through Ruby may be cached regardless of the uncacheable flag, while accesses to IO/device memory bypass the Ruby system and are never cached. This due to the way the ruby sequencer converts gem5 packets to RubyRequest messages and is actually not specific to CHI.

A few ways you could work around this:

- set the alloc_on_* parameters of CHI caches to false, so nothing is ever cached
- or modify the Sequencer to issue different requests for uncacheable accesses and have CHI-cache handle them accordingly.

**Reply by hnpl** [reply]
Thanks @tiagormk! That's really helpful.
